taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/service/CamundaTaskEventsService.java

```diff
@@ -116,6 +143,224 @@ public void deleteEvents(String idsAsJsonArray) {
     }
   }
 
+  public void decreaseRemainingRetriesAndLogError(String eventIdAndErrorLog) {
+
+    String sql = String.format(SQL_DECREASE_REMAINING_RETRIES, OUTBOX_SCHEMA);
+
+    try (Connection connection = getConnection();
+        PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
+
+      JsonNode id = objectMapper.readTree(eventIdAndErrorLog).get("taskEventId");
+
+      JsonNode errorLog = objectMapper.readTree(eventIdAndErrorLog).get("errorLog");
+
+      Instant blockedUntil = getBlockedUntil();
+
+      preparedStatement.setTimestamp(1, Timestamp.from(blockedUntil));
+      preparedStatement.setString(2, errorLog.asText());
+      preparedStatement.setInt(3, id.asInt());
+      preparedStatement.execute();
+
+    } catch (Exception e) {
+      LOGGER.warn(
+          "Caught Exception while trying to decrease the remaining retries of camunda task event",
+          e);
+    }
+  }
+
+  public List<CamundaTaskEvent> getEventsFilteredByRetries(Integer remainingRetries) {
+
+    List<CamundaTaskEvent> camundaTaskEventsFilteredByRetries = new ArrayList<>();
+
+    String getEventsFilteredByRetriesSql =
+        String.format(SQL_GET_EVENTS_FILTERED_BY_RETRIES, OUTBOX_SCHEMA);
+
+    try (Connection connection = getConnection();
+        PreparedStatement preparedStatement =
+            connection.prepareStatement(getEventsFilteredByRetriesSql)) {
+
+      preparedStatement.setInt(1, remainingRetries);
+
+      ResultSet camundaTaskEventFilteredByRetriesResultSet = preparedStatement.executeQuery();
+      camundaTaskEventsFilteredByRetries =
+          getCamundaTaskEvents(camundaTaskEventFilteredByRetriesResultSet);
+
+    } catch (Exception e) {
+      LOGGER.warn("Caught Exception while trying to retrieve failed events from the outbox", e);
+    }
+    return camundaTaskEventsFilteredByRetries;
+  }
+
+  public String getEventsCount(int remainingRetries) {
+
+    String eventsCount = "{\"eventsCount\":0}";
+
+    String getEventsCountSql = String.format(SQL_GET_EVENTS_COUNT, OUTBOX_SCHEMA);
+
+    try (Connection connection = getConnection();
+        PreparedStatement preparedStatement = connection.prepareStatement(getEventsCountSql)) {
+
+      preparedStatement.setInt(1, remainingRetries);
+
+      ResultSet camundaTaskEventResultSet = preparedStatement.executeQuery();
+
+      if (camundaTaskEventResultSet.next()) {
+
+        eventsCount = eventsCount.replace("0", String.valueOf(camundaTaskEventResultSet.getInt(1)));
+      }
+
+    } catch (Exception e) {
+      LOGGER.warn("Caught Exception while trying to retrieve events count from the outbox", e);
+    }
+    return eventsCount;
+  }
+
+  public CamundaTaskEvent setRemainingRetries(int id, Map<String, Integer> newRemainingRetries) {
```

Comment:
I personally don't like to use the Map in the signature of this operation. Why not move the Map handling to the controller and specify the number of retries here?

Reply:
Changed